compiler/noirc_evaluator/src/ssa/opt/basic_conditional.rs

```diff
@@ -0,0 +1,522 @@
+        dfg::DataFlowGraph,
+        function::{Function, FunctionId},
+        function_inserter::FunctionInserter,
+        instruction::{BinaryOp, Instruction, TerminatorInstruction},
+        post_order::PostOrder,
+        value::ValueId,
+    },
+    Ssa,
+};
+
+use super::flatten_cfg::Context;
+#[derive(Debug, Clone)]
+struct BasicConditional {
+    block_entry: BasicBlockId,
+    block_then: Option<BasicBlockId>,
+    block_else: Option<BasicBlockId>,
+    block_exit: BasicBlockId,
+}
+
+impl Ssa {
+    #[tracing::instrument(level = "trace", skip(self))]
+    pub(crate) fn flatten_basic_conditionals(mut self) -> Ssa {
+        // Retrieve the 'no_predicates' attribute of the functions in a map, to avoid problems with borrowing
+        let mut no_predicates = HashMap::default();
+        for function in self.functions.values() {
+            no_predicates.insert(function.id(), function.is_no_predicates());
+        }
+        for function in self.functions.values_mut() {
+            flatten_function(function, &mut no_predicates);
+        }
+        self
+    }
+}
+
+// Returns the blocks of the simple conditional sub-graph whose input block is the entry.
+// Returns None if the input block is not the entry block of a simple conditional.
+fn is_conditional(
+    block: BasicBlockId,
+    cfg: &ControlFlowGraph,
+    function: &Function,
+) -> Option<BasicConditional> {
+    // jump overhead is the cost for doing the conditional and jump around the blocks
+    // We use 10 as a rough estimate, the real cost is less.
+    let jump_overhead = 10;
+    let mut successors = cfg.successors(block);
+    let mut result = None;
+    // a conditional must have 2 branches
+    if successors.len() != 2 {
+        return None;
+    }
+    let left = successors.next().unwrap();
+    let right = successors.next().unwrap();
+    let mut left_successors = cfg.successors(left);
+    let mut right_successors = cfg.successors(right);
+    let left_successors_len = left_successors.len();
+    let right_successors_len = right_successors.len();
+    let next_left = left_successors.next();
+    let next_right = right_successors.next();
+    if next_left == Some(block) || next_right == Some(block) {
+        // this is a loop, not a conditional
+        return None;
+    }
+    if left_successors_len == 1 && right_successors_len == 1 && next_left == next_right {
+        // The branches join on one block so it is a non-nested conditional
+        let cost_left = block_cost(left, &function.dfg);
+        let cost_right = block_cost(right, &function.dfg);
+        // For the flattening to be valuable, we compare the cost of the flattened code with the average cost of the 2 branches,
+        // including an overhead to take into account the jumps between the blocks.
+        let cost = cost_right.saturating_add(cost_left);
+        if cost < cost / 2 + jump_overhead {
+            if let Some(TerminatorInstruction::JmpIf {
+                condition: _,
```

Comment:
What happens to the `condition`? Based on the ticket description I thought it would be come a multiplying factor in the merging of the two blocks somehow.

Reply:
I see from the tests that the multiplication indeed happens, I guess somehow facilitated by the `Context`. I suppose one has to be super familiar with the `flatten_cfg` to understand where the magic happens.